### PR TITLE
Wrap unsafe libltdl calls with locking

### PR DESCRIPTION
**Pull request by msmeissn**
https://github.com/gphoto/libgphoto2/issues/166

needs review

### CODE REVIEW

File: libgphoto2_port/Makefile.am [hunk 1]
```diff
@@ -96,6 +96,7 @@ noinst_DATA = libgphoto2_port-uninstalled.pc
 
 # ship headers
 nobase_include_HEADERS =	\
+	gphoto2/gphoto2-port-mutex.h			\
```
Comment (ndim) [inline libgphoto2_port/Makefile.am:99]: Why install an internal header file (`gpi_*`)?

**Review by ndim — COMMENTED** [review]
I will have to examine this on my development workstation as well, not just on my phone.

File: libgphoto2_port/libgphoto2_port/Makefile.am [hunk 2]
```diff
@@ -62,15 +62,15 @@ libgphoto2_port_la_LDFLAGS      += -version-info @LIBGPHOTO2_PORT_VERSION_INFO@
 libgphoto2_port_la_CPPFLAGS     += $(LTDLINCL)
 libgphoto2_port_la_LIBADD       += $(LIBLTDL)
 
-libgphoto2_port_la_LIBADD       += $(INTLLIBS)
```
[the remark this answers is not in the document]
Comment (msmeissn) [reply libgphoto2_port/libgphoto2_port/Makefile.am:65]: merge issue. removed

File: libgphoto2/gphoto2-abilities-list.c [hunk 3]
```diff
@@ -212,10 +213,12 @@ gp_abilities_list_load_dir (CameraAbilitiesList *list, const char *dir,
 	if (1) { /* a new block in which we can define a temporary variable */
 		foreach_data_t foreach_data = { NULL, GP_OK };
 		foreach_data.list = flist;
+		pthread_mutex_lock(&gpi_libltdl_mutex);
```
[the remark this answers is not in the document]
Comment (msmeissn) [reply libgphoto2/gphoto2-abilities-list.c:216]: we are so far not using pthread at all. 
we could put into 2 specific functions, yes

File: libgphoto2_port/Makefile.am (hunk 1 above)
Comment (msmeissn) [reply libgphoto2_port/Makefile.am:99]: hmm, we export it for libgphoto2 usage. but probably not needed there

**Comment by ndim** [comment]
@msmeissn Do you mind if I push a few commits resolving the remarks I have made above, and then we can both take a look at the results?

**Comment by ndim** [comment]
I have a pushed a few single commits for easier review, but recommend squashing the commit series into a single commit before committing to master.

**Comment by ndim** [comment]
Two issues remaining:

- [ ] Is the copyright line `Copyright 2017 kadler15 <spurfan15@gmail.com>` good?
- [ ] Is having more than a single `lt_*()' function call in a mutex locked region, required, problematic, or just lazy?

**Comment by msmeissn** [comment]
> @msmeissn Do you mind if I push a few commits resolving the remarks I have made above, and then we can both take a look at the results?

just go ahead, thanks. :) (I am currently sick and otherwise also busy on most days... so expect some delays from me.)

**Comment by msmeissn** [comment]
I think abstracting this into 2 functions and not exposing the pthread mutex over the library is fine. 

one thing, is lt_dlclose / lt_dlopenext refcounting? otherwise we could still load/unload code from other threads.
(e.g. think:
- one thread operating already in a ptp driver driven camera
- second thread does autodetection, dlclose//dlopens around including ptp2 driver
will this interact badly?)

**Comment by ndim** [comment]
I am starting to think we should just seriously look into statically linking all the camlibs into the library. :-/

And do get well soon.

**Comment by ndim** [comment]
> one thing, is lt_dlclose / lt_dlopenext refcounting? otherwise we could still load/unload code from other threads. (e.g. think:
> 
>     * one thread operating already in a ptp driver driven camera
> 
>     * second thread does autodetection, dlclose//dlopens around including ptp2 driver
>       will this interact badly?)

Quoting the libtool info page:

>  -- Function: int lt_dlclose (lt_dlhandle HANDLE)
     Decrement the reference count on the module HANDLE.  If it drops to
     zero and no other module depends on this module, then the module is
     unloaded.  Return 0 on success.

Looks like reference counting to me, and should alleviate the concerns you express above.

**Comment by ndim** [comment]
So, code-wise this looks good to me right now.

The question of the copyright notices remains. The modified files have (according to the git history) about twenty authors. Two or three are mentioned in the copyright notice. Just adding original locking author @kadler15 makes them between a quarter and a half of the authors mentioned in the copyright comment, which looks a bit unfair to all the other authors not mentioned.

So I can either

  * First convert the git history into accurate copyright information (from `Author:`, `Committer:`, and possibly a third person mentioned from the commit message itself), and then add kadler15, and then merge this. This is a lot of work.
  * Just add kadler15 to the modified files, and then merge this, ignoring the other authors. Not that much work, but inconsistent.
  * Just merge this without adding any more copyright notices. Also not much work, also inconsistent.

The latter two options mean kicking the can down the road some more. However, I would want a bit of scripting support to adapt *all* source files' comment in the git repo.

File: libgphoto2/gphoto2-abilities-list.c [hunk 4]
```diff
@@ -212,10 +213,12 @@ gp_abilities_list_load_dir (CameraAbilitiesList *list, const char *dir,
 	if (1) { /* a new block in which we can define a temporary variable */
 		foreach_data_t foreach_data = { NULL, GP_OK };
 		foreach_data.list = flist;
+		gpi_libltdl_lock();
```
Comment (RReverser) [inline libgphoto2/gphoto2-abilities-list.c:216]: Looking at frequent locks/unlocks throughout the function, I wonder if it would've been easier to

1. Rename the unmodified function to `gp_abilities_list_load_dir_impl`.
2. Add a function `gp_abilities_list_load_dir` that locks libltdl, invokes `gp_abilities_list_load_dir_impl`, unlocks libltdl and then returns the return code.

Comment (RReverser) [reply libgphoto2/gphoto2-abilities-list.c:216]: Same for other functions that are dedicated to loading libs, and consist for ~90% of libltdl calls anyway.

**Comment by ndim** [comment]
JFTR, regarding the copyright notices: https://github.com/gphoto/libgphoto2/issues/851

File: libgphoto2/gphoto2-abilities-list.c (hunk 4 above)
Comment (msmeissn) [reply libgphoto2/gphoto2-abilities-list.c:216]: @ndim what do you think? seems sensible

File: libgphoto2_port/libgphoto2_port/Makefile.am [hunk 5]
```diff
@@ -62,15 +62,17 @@ libgphoto2_port_la_LDFLAGS      += -version-info @LIBGPHOTO2_PORT_VERSION_INFO@
...
+libgphoto2_port_la_LIBADD       += $(INTLLIBS)
```
Comment (Sija) [inline libgphoto2_port/libgphoto2_port/Makefile.am:65]: Unrelated change?

**Comment by msmeissn** [comment]
hmm, some drivers might not be threadsafe, but generally i would hope we are.

**Comment by RReverser** [comment]
> hmm, some drivers might not be threadsafe, but generally i would hope we are.

I found a lot of static mutable variables, as well as MT-unsafe C function calls. I started fixing some of them, but there's too much work so gave up - if it helps, I can send a WIP PR together with Clang analysis if someone wants to take over.

**Comment by ndim** [comment]
@RReverser If you could reply with a few helpful example clang analysis commands to start with (for someone unfamiliar with the tool), that would already go a long way.

**Comment by RReverser** [comment]
By default it provides a lot of noise (false positives), let me filter them out a little more and then I'll share. Most problematic code is indeed in various camera drivers, but then, those are inherent to libgphoto2, so to fix thread safety they would need to be fixed too.

I'll share something in a few days.

File: libgphoto2/gphoto2-abilities-list.c (hunk 4 above)
Comment (RReverser) [reply libgphoto2/gphoto2-abilities-list.c:216]: While I'm working on other parts, let's do this anyway ^

File: libgphoto2_port/libgphoto2_port/Makefile.am (hunk 5 above)
Comment (msmeissn) [reply libgphoto2_port/libgphoto2_port/Makefile.am]: yes. removed

File: libgphoto2/gphoto2-abilities-list.c (hunk 4 above)
Comment (msmeissn) [reply libgphoto2/gphoto2-abilities-list.c]: i did this for this function.
